flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java

```diff
@@ -0,0 +1,830 @@
+        super.close();
+    }
+
+    public void processElement(int inputId, StreamRecord<RowData> element) throws Exception {
+        RowData input = element.getValue();
+        if (input == null) {
+            return;
+        }
+
+        performMultiJoin(input, inputId);
+        addRecordToState(input, inputId);
+    }
+
+    private void performMultiJoin(RowData input, int inputId) throws Exception {
+        int[] associations = createInitialAssociations();
+        RowData[] currentRows = new RowData[inputSpecs.size()];
+
+        recursiveMultiJoin(
+                0, input, inputId, currentRows, associations, JoinPhase.CALCULATE_MATCHES);
+    }
+
+    /**
+     * See {@link StreamingMultiJoinOperator} for a detailed explanation of the recursive join and
+     * examples.
+     *
+     * @param depth The current depth of the recursion, representing the input stream index (0 to
+     *     N-1).
+     * @param input The original input record that triggered this join operation.
+     * @param inputId The index of the input stream from which the `input` record originated.
+     * @param currentRows An array holding the candidate row from each input stream processed so far
+     *     in this recursive path. `currentRows[d]` holds the row from input `d`.
+     * @param associations An array used for LEFT joins to track match counts. `associations[d]`
+     *     stores the number of successful matches found for `currentRows[d]` against inputs `d+1`
+     *     onwards based on outer join conditions.
+     * @param phase The current execution phase (CALCULATE_MATCHES or EMIT_RESULTS).
+     * @throws Exception If state access or condition evaluation fails.
+     */
+    private void recursiveMultiJoin(
+            int depth,
+            RowData input,
+            int inputId,
+            RowData[] currentRows,
+            int[] associations,
+            JoinPhase phase)
+            throws Exception {
+        // Base case: If we've processed all inputs, evaluate the final join condition.
+        if (isMaxDepth(depth)) {
+            emitJoinedRow(input, currentRows);
+            return;
+        }
+        boolean isLeftJoin = isLeftJoinAtDepth(depth);
+
+        // Store a record from the current level in currentRows and go to the next index.
+        // We iterate through records using a recursive depth first search approach.
+        // We store matched here because we need to know if we need to do emit a null padded output
+        // if there were no matching records.
+        boolean matched =
+                processRecords(depth, input, inputId, currentRows, associations, phase, isLeftJoin);
+
+        // If the current depth is the one where the triggering input record arrived,
+        // now process the input record itself with the current combination of rows we are at.
+        if (isInputLevel(depth, inputId)) {
+            processInputRecord(depth, input, inputId, currentRows, associations, matched);
```

Comment:
Am I correct that we call this method `processInputRecord` only here for condition
```java
if (isInputLevel(depth, inputId))
```

then the question if we need to pass both `depth` and `inputId`?

Asking since based on body of `isInputLevel`
```java
    private boolean isInputLevel(int depth, int inputId) {
        return depth == inputId;
    }
```
we can drop one of them from arg list

Reply:
Also a question: I wonder if we need `isInputLevel` method at all?
Asking since I notice that there is a number of places where we have same conditions in `if` without usage of this method

Reply:
> Asking since I notice that there is a number of places where we have same conditions in if without usage of this method

Oh yeah, true. I've refactored, so the other two places use it. Let me know if I forgot one

> we can drop one of them from arg list
What could we drop? I didn't get this part